tsdb/head_wal.go

```diff
@@ -426,15 +413,23 @@ func (wp *walSubsetProcessor) reuseBuf() []record.RefSample {
 // processWALSamples adds the samples it receives to the head and passes
 // the buffer received to an output channel for reuse.
 // Samples before the minValidTime timestamp are discarded.
-func (wp *walSubsetProcessor) processWALSamples(h *Head) (unknownRefs uint64) {
+func (wp *walSubsetProcessor) processWALSamples(h *Head, mmappedChunks map[chunks.HeadSeriesRef][]*mmappedChunk) (unknownRefs, mmapOverlappingChunks uint64) {
 	defer close(wp.output)
 
 	minValidTime := h.minValidTime.Load()
 	mint, maxt := int64(math.MaxInt64), int64(math.MinInt64)
 
-	for samples := range wp.input {
-		wp.mx.Lock()
-		for _, s := range samples {
+	for in := range wp.input {
+		if in.oldSeriesRef != 0 && in.newSeriesRef != 0 {
+			mSeries := h.series.getByID(in.oldSeriesRef)
```

Comment:
Why don't we just pass the `oldSeries *memSeries` in the `walSubsetProcessorInputItem`? We already have it when we enqueue this item, we pass their reference, and then we `get` them again.

Reply:
+1